text/0000-any_type_self.md

```diff
@@ -0,0 +1,79 @@
+- Feature Name: any_type_self
+- Start Date: 2016-07-26
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow self parameter be any type.
+
+# Motivation
+[motivation]: #motivation
+
+It allows to avoid wrapper structs.
+
+# Detailed design
+[design]: #detailed-design
+
+Currently self parameter type is allowed to be only `Self`, `&Self`, `&mut Self`, `Box<Self>`.
+
+Solution: Allow `self` be any type that depends on `Self`.
```

Comment:
@ticki I can't understand you.
*Type* depends on `Self` = *Type* has `Self` as generic parameter or associated type.